### PR TITLE
Hitbox now with Circle Hitboxes + Tests

### PR DESCRIPTION
**Pull request by gordon-cheung**
This change is [Reviewable](https://reviewable.io/reviews/enochtsang/terminalfighter/68)

### CODE REVIEW

**Comment by ExtraGravity** [comment]
I linted the files using the AStyleFormatter, this adjusts the look of the code to make sure it looks nice, you should look into getting a linter tool yourself.

**Comment by ExtraGravity** [comment]
Pretty good overall! Very proud of you gordo. What I would like to comment on is that you're using `0` to check that it's not a circle or not a hitbox, this isn't very safe because it relies on a developer never putting in a 0. What I suggest instead is to keep a private member that will define whether it's a circle or rectangle. 

I would do this through an enum (you can look that up), you will have an access function called `type` and you'd return an enum type `RECT` or `CIRCLE`.  Then in your `is_overlapping()` function you can use that to check for the type instead of checking for `0`s.

File: src/hitbox.h [hunk 1]
```diff
@@ -1,13 +1,29 @@
...
+class Hitbox {
+public:
+    Hitbox(double x_pos, double y_pos, double width, double height);
+    Hitbox(double x_pos, double y_pos, double radius);
+    Hitbox(SDL_Rect hitbox);
+    Hitbox(Circle hitbox);
+
+    bool is_overlapping(Hitbox& other_hitbox);
+    const SDL_Rect& hitbox() const;
+    const Circle& hitbox_circle() const;
```
Comment (ExtraGravity) [inline src/hitbox.h]: I'd recommend returning pointers here instead, that way you can return a NULL value instead, returning an actual value that may not be valid is unsafe.

File: src/menu_state.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+public:
+	MenuState(SDL_Renderer* renderer);
+    ~MenuState();
+	gamestates::GameStateName run();
+	gamestates::GameStateName name() const;
+
+private:
+	void display_debug_frames(Delay *delayer);
+	SDL_Renderer* renderer_;
+};
```
Comment (ExtraGravity) [inline src/menu_state.h]: Lol why is this file in this PR haha.

**Comment by ExtraGravity** [comment]
I'll leave it here for now.

**Comment by gordon-cheung** [comment]
Thanks Enoch, I'll try to get it done by Sunday, Jan 31. I have no idea why the menu state is in there, possibly because I merged master to this branch?

**Comment by bseto** [comment]
When grabbing master into your feature branch, we should start using the command:
`git rebase origin/master`. This will make it a rewind and place you ontop of master rather then "merging"

**Comment by ExtraGravity** [comment]
Why do you say that in this PR?

**Comment by bseto** [comment]
Cause it has to do with the "Lol why is this file in this PR haha."

**Comment by ExtraGravity** [comment]
Ah makes sense, that would solve that issue.

**Comment by ExtraGravity** [comment]
Review status: 0 of 3 files reviewed at latest revision, 17 unresolved discussions.

---

_[src/hitbox.cpp, line 9 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1gsJ0K66MItsrcSYW:-KS1gsJ0K66MItsrcSYX:bdw1n7z) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L9)):_

> ``` cpp
>     hitbox_.h = height;
> 
>     hitbox_type_ = rect;
> ```

Should explicitly the circle_hitbox to NULL, and i bbelieve we can still use inline construction for the hitbox. 

`: hitbox_(x, y, width, height)` should still work I think.

---

_[src/hitbox.cpp, line 14 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1hwyKFvaKy0FSquei:-KS1hwyKFvaKy0FSquej:b2r30ir) ([raw file](https://github.com/enochtsang/terminalfighter/blob/0807983870db956a198545488db9381ea5787208/src/hitbox.cpp#L14)):_

> ``` cpp
> }
> 
> bool Hitbox::is_overlapping(const Hitbox& other_hitbox) {
> ```

What was wrong with const? I believe we should still be using const.

---

_[src/hitbox.cpp, line 15 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1gjQNkoRgh0J_GGys:-KS1gjQNkoRgh0J_GGyt:bng0fun) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L15)):_

> ``` cpp
>     circle_hitbox_.x = x_pos;
>     circle_hitbox_.y = y_pos;
>     circle_hitbox_.r = radius;
> ```

Same as above, we should use inline construction if we can, `hitbox_type_` for sure we can use it.

---

_[src/hitbox.cpp, line 21 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1i06p8wVIUor_hH9M:-KS1i06p8wVIUor_hH9N:be89wt3) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L21)):_

> ``` cpp
> 
> Hitbox::Hitbox(SDL_Rect hitbox) {
>     hitbox_ = hitbox;
> ```

Set circle hitbox to NULL, set hitbox_type_ and use inline construction.

---

_[src/hitbox.cpp, line 25 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1iLJVxRKuw4rmKjRk:-KS1iLJVxRKuw4rmKjRl:b-4h98sp) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L25)):_

> ``` cpp
> 
> Hitbox::Hitbox(Circle hitbox) {
>     circle_hitbox_ = hitbox;
> ```

Set hitbox_type_ and use inline construction.

---

_[src/hitbox.cpp, line 30 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1hU1cwoVFvHsfFLYT:-KS1hU1cwoVFvHsfFLYU:b-c2rlxe) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L30)):_

> ``` cpp
> 
> bool Hitbox::is_overlapping(Hitbox& other_hitbox) {
>     bool is_overlapping;
> ```

instead of keeping the variable, can we just straight up return the results.

---

_[src/hitbox.cpp, line 31 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1i8BS6YQtCTXp6rWY:-KS1i8BS6YQtCTXp6rWZ:b-2kwz1c) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L31)):_

> ``` cpp
> bool Hitbox::is_overlapping(Hitbox& other_hitbox) {
>     bool is_overlapping;
> ```

What do you think of using an assert here to check that at least one of circle_hitbox_ or hitbox_ is null depeneding on the hitbox_type value.

---

_[src/hitbox.cpp, line 39 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1esxGG7T_0r_Vvls4:-KS1esxGG7T_0r_Vvls5:brdk0gx) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L39)):_

> ``` cpp
>         is_overlapping = check_overlap(this->hitbox_, other_hitbox.hitbox_);
>     // CHECKS COLLISION BETWEEN CIRCLE AND RECT
>     else if (*(this->hitbox_type()) == circle && *(other_hitbox.hitbox_type()) == rect)
> ```

We could just use one else statement. So if 2 circles, else if 2 boxes, else everything else.

---

_[src/hitbox.cpp, line 42 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1e3JbyyJHktGWeTuN:-KS1e3JbyyJHktGWeTuO:b-g8wogi) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L42)):_

> ``` cpp
>         is_overlapping = check_overlap(this->circle_hitbox_, other_hitbox.hitbox_);
>     // CHECKS COLLISION BETWEEN RECT AND CIRCLE
>     else if (*(this->hitbox_type()) == rect && *(other_hitbox.hitbox_type()) == circle)
> ```

Braces around if else statements.

---

_[src/hitbox.cpp, line 59 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1jiIb2ODGJXtTBCI0:-KS1jiIb2ODGJXtTBCI1:bupeobh) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L59)):_

> ``` cpp
>     if (distanceSquared <= radiusSquared)
>     {
>         return true;
> ```

We can do `return distanceSquared <= radiusSquared` for less code.

---

_[src/hitbox.cpp, line 79 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1iSG-6J96TgxRhVbB:-KS1iSG-6J96TgxRhVbC:b-g8wogi) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L79)):_

> ``` cpp
>         return false;
> 
>     if (distance_x <= (other_hitbox.w / 2))
> ```

Braces around if else statements.

---

_[src/hitbox.cpp, line 85 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1jxPzi0w0OWN5KRvF:-KS1jxPzi0w0OWN5KRvG:b-fue9gt) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L85)):_

> ``` cpp
>         return true;
> 
>     double d_x = (distance_x - other_hitbox.w / 2);
> ```

`d_x` looks a bit funky, I think `dx` is will look better but it doesn't really matter, just up for discussion.

---

_[src/hitbox.cpp, line 90 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1k6MvbT6RrTdd_F1T:-KS1k6MvbT6RrTdd_F1U:b-cqkpap) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L90)):_

> ``` cpp
>     d_y = d_y * d_y;
> 
>     double cornerDistance_sq = d_x + d_y;
> ```

I think it makes more sense to do the squaring here, because `d_y = d_y * d_y` doesn't keep `d_y` as actually `d_y`.

---

_[src/hitbox.h, line 22 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1kJ5mds3uRcit2ngG:-KS1kJ5mds3uRcit2ngH:b-x2y12e) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.h#L22)):_

> ``` objc
>         const SDL_Rect* hitbox() const;
>         const Circle* hitbox_circle() const;
>         const Shape* hitbox_type() const;
> ```

It's just an enum which evaluates to just an number in memory, so I think we can return pass by value for less complexity without losing performance.

---

_[src/hitbox.h, line 27 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1kYPd5iensTi4Ctv_:-KS1kYPd5iensTi4Ctva:b-t1vi98) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.h#L27)):_

> ``` objc
>         SDL_Rect hitbox_;
>         Circle circle_hitbox_;
>         Shape hitbox_type_;
> ```

Why are some protected and private? I feel like we must've had a good reason before, but I don't know what.

---

_[test/src/hitbox_test.cpp, line 53 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1gbTrFOXYdwsQw9rn:-KS1gbTrFOXYdwsQw9ro:b3f7b4b) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/test/src/hitbox_test.cpp#L53)):_

> ``` cpp
> }
> 
> TEST (HitboxTest, circle_member_constructor) {
> ```

The tests are great, but can we organize them in a way that makes more sense?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68)_

**Comment by bseto** [comment]
Review status: 0 of 3 files reviewed at latest revision, 16 unresolved discussions.

---

_[src/hitbox.cpp, line 9 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1gsJ0K66MItsrcSYW:-KS3IBP_YzSa9mbtH13-:br3yogv) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L9)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> Should explicitly the circle_hitbox to NULL, and i bbelieve we can still use inline construction for the hitbox. 
> 
> `: hitbox_(x, y, width, height)` should still work I think.
> </details>
> circle_hitbox isn't a pointer though. I think we can just leave it as is as long as they check for hitbox_type_ it should be fine. 

---

_[src/hitbox.cpp, line 14 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1hwyKFvaKy0FSquei:-KS2yvqvpQTOLVGx7IYv:b32w0x) ([raw file](https://github.com/enochtsang/terminalfighter/blob/0807983870db956a198545488db9381ea5787208/src/hitbox.cpp#L14)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> What was wrong with const? I believe we should still be using const.
> </details>
> OK

---

_[src/hitbox.cpp, line 21 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1i06p8wVIUor_hH9M:-KS3QfWo44lw-v4TpG-d:b-kyq5ft) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L21)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> Set circle hitbox to NULL, set hitbox_type_ and use inline construction.
> </details>
> Setting hitbox_type_ inline construction, but not making circle_hitbox to NULL since it's not a pointer. 

---

_[src/hitbox.cpp, line 30 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1hU1cwoVFvHsfFLYT:-KS3SABwgB4K4Imi-mOU:b94da08) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L30)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> instead of keeping the variable, can we just straight up return the results.
> </details>
> Would you rather have a return on every if else? I think it looks good on how it is right now. 

---

_[src/hitbox.cpp, line 31 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1i8BS6YQtCTXp6rWY:-KS3TRHcomw1IOz3fBDZ:b-asu5c0) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L31)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> What do you think of using an assert here to check that at least one of circle_hitbox_ or hitbox_ is null depeneding on the hitbox_type value.
> </details>

Well, because they aren't pointers, they won't be NULL. I don't think we need to check it.

---

_[src/hitbox.cpp, line 39 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1esxGG7T_0r_Vvls4:-KS3YncTligSOFp89DsR:bowz7ii) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L39)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> We could just use one else statement. So if 2 circles, else if 2 boxes, else everything else.
> </details>
> I think it has to be:

If 2 circles
else if 2 boxes
else if (circle and box) *_Order matters
else (box and circle) *_Order matters

We can't just "else" everything else because the order matters.

---

_[src/hitbox.cpp, line 42 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1e3JbyyJHktGWeTuN:-KS3b-ExCpV4K1xGySEH:b32w0x) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L42)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> Braces around if else statements.
> </details>
> OK

---

_[src/hitbox.cpp, line 59 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1jiIb2ODGJXtTBCI0:-KS3bCenNsc-f9FcvGv8:b2nhksp) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L59)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> We can do `return distanceSquared <= radiusSquared` for less code.
> </details>
> OK

---

_[src/hitbox.cpp, line 79 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1iSG-6J96TgxRhVbB:-KS3bLpdgLCjO30vydmB:b2nhksp) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L79)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> Braces around if else statements.
> </details>
> OK

---

_[src/hitbox.cpp, line 85 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1jxPzi0w0OWN5KRvF:-KS3bYjjog0uPv3VQ2MX:b-7iyuyi) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L85)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> `d_x` looks a bit funky, I think `dx` is will look better but it doesn't really matter, just up for discussion.
> </details>
> looks kinda funky but I think it illustrates the _x is a value instead of dx being one thing a bit better. I'll just leave it.

---

_[src/hitbox.cpp, line 90 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1k6MvbT6RrTdd_F1T:-KS3c1-zvqxFNjTTuKr1:b-3mmbse) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L90)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> I think it makes more sense to do the squaring here, because `d_y = d_y * d_y` doesn't keep `d_y` as actually `d_y`.
> </details>
> OK added the squaring to be on that line and added parenthesis just for clarity. 

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68)_

**Comment by ExtraGravity** [comment]
Reviewed 4 of 4 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

_[src/hitbox.cpp, line 9 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1gsJ0K66MItsrcSYW:-KS50mmPGIm6i9bs-7P7:b-ywq9ko) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L9)):_
<details><summary>

<i>Previously, bseto (Byron Seto) wrote…</i></summary>



> circle_hitbox isn't a pointer though. I think we can just leave it as is as long as they check for hitbox_type_ it should be fine. 
> </details>
> I'm pretty sure we can use inline construction for `hitbox_` as per the above comment.

---

_[src/hitbox.cpp, line 15 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1gjQNkoRgh0J_GGys:-KS50vhS-rxma5s2XE3D:boah2i2) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L15)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> Same as above, we should use inline construction if we can, `hitbox_type_` for sure we can use it.
> </details>
> Pretty sure you can use inline consturction for circle_hitbox_. Might have to add the constructor for circle.

---

_[src/hitbox.cpp, line 21 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1i06p8wVIUor_hH9M:-KS51Cl5ojQB3fqoo4-2:b-9eyl5a) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L21)):_
<details><summary>

<i>Previously, bseto (Byron Seto) wrote…</i></summary>



> Setting hitbox_type_ inline construction, but not making circle_hitbox to NULL since it's not a pointer. 
> </details>
> I don't see `hitbox_type_` being set.

---

_[src/hitbox.cpp, line 25 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1iLJVxRKuw4rmKjRk:-KS51L3D2t-cpCa8uYKs:b-9eyl5a) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L25)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> Set hitbox_type_ and use inline construction.
> </details>
> I don't see `hitbox_type_` being set.

---

_[src/hitbox.cpp, line 30 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1hU1cwoVFvHsfFLYT:-KS51r0g8mwZZMMMh9kJ:bjfzcl) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L30)):_
<details><summary>

<i>Previously, bseto (Byron Seto) wrote…</i></summary>



> Would you rather have a return on every if else? I think it looks good on how it is right now. 
> </details>
> Sounds good, you need to set this as resolved yourself.

---

_[src/hitbox.cpp, line 31 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1i8BS6YQtCTXp6rWY:-KS51rdP_XHXtYLx5qFI:bjfzcl) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L31)):_
<details><summary>

<i>Previously, bseto (Byron Seto) wrote…</i></summary>



> Well, because they aren't pointers, they won't be NULL. I don't think we need to check it.
> </details>
> Sounds good, you need to set this as resolved yourself.

---

_[src/hitbox.cpp, line 39 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1esxGG7T_0r_Vvls4:-KS51igKyAcCw2nsxC1R:bjfzcl) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L39)):_
<details><summary>

<i>Previously, bseto (Byron Seto) wrote…</i></summary>



> I think it has to be:
> 
> If 2 circles
> else if 2 boxes
> else if (circle and box) *_Order matters
> else (box and circle) *_Order matters
> 
> We can't just "else" everything else because the order matters.
> 
> </details>
> Sounds good, you need to set this as resolved yourself.

---

_[src/hitbox.cpp, line 85 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1jxPzi0w0OWN5KRvF:-KS522OWPSrGbF6udGEZ:bjfzcl) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L85)):_
<details><summary>

<i>Previously, bseto (Byron Seto) wrote…</i></summary>



> looks kinda funky but I think it illustrates the _x is a value instead of dx being one thing a bit better. I'll just leave it.
> </details>
> Sounds good, you need to set this as resolved yourself.

---

_[src/hitbox.cpp, line 90 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1k6MvbT6RrTdd_F1T:-KS525zMDBX01IOm7qKd:bsrfkwe) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L90)):_
<details><summary>

<i>Previously, bseto (Byron Seto) wrote…</i></summary>



> OK added the squaring to be on that line and added parenthesis just for clarity. 
> </details>
> Looks good

---

_[src/hitbox.h, line 27 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1kYPd5iensTi4Ctv_:-KS52HWUkZToypfpDSfM:b-5ca9mg) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.h#L27)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> Why are some protected and private? I feel like we must've had a good reason before, but I don't know what.
> </details>
> The indentation is weird now, but still why are some protected vs private?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68)_

**Comment by bseto** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

_[src/hitbox.cpp, line 9 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1gsJ0K66MItsrcSYW:-KS5fjMUnTrN9Uf7E6Im:by94j1s) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L9)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> I'm pretty sure we can use inline construction for `hitbox_` as per the above comment.
> </details>
> How?

`:hitbox_.x(x_pos),
hitbox_.y(ypos),
hitbox_.w(width),
hitbox_.h(height)`
didn't work when i tried it earlier.  I'll try it again in the morning, but what is your suggestion?

And doing:

`Hitbox::Hitbox(double x_pos, double y_pos, double width, double height)
:hitbox_(xpos, y_pos, width, height)`

Won't work either. It will just infinitely call itself right? 

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68)_

**Comment by ExtraGravity** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

_[src/hitbox.cpp, line 9 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1gsJ0K66MItsrcSYW:-KS6uDXlroZO9Y4CUNje:bi3xh1a) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L9)):_
<details><summary>

<i>Previously, bseto (Byron Seto) wrote…</i></summary>



> How?
> 
> `:hitbox_.x(x_pos),
> hitbox_.y(ypos),
> hitbox_.w(width),
> hitbox_.h(height)`
> didn't work when i tried it earlier.  I'll try it again in the morning, but what is your suggestion?
> 
> And doing:
> 
> `Hitbox::Hitbox(double x_pos, double y_pos, double width, double height)
> :hitbox_(xpos, y_pos, width, height)`
> 
> Won't work either. It will just infinitely call itself right? 
> 
> </details>
> the second one will work, because it's not calling the Hitbox constructor, it's calling the SDL_Rect constructor, as per it's type.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68)_

**Comment by bseto** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

_[src/hitbox.cpp, line 9 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1gsJ0K66MItsrcSYW:-KS7sIKMjv7LAYh4qbRF:b32w0x) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L9)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> the second one will work, because it's not calling the Hitbox constructor, it's calling the SDL_Rect constructor, as per it's type.
> </details>
> OK

---

_[src/hitbox.h, line 27 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1kYPd5iensTi4Ctv_:-KS8nvAFyXzdexRnobca:bim3iyu) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.h#L27)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> The indentation is weird now, but still why are some protected vs private?
> </details>
> I'm actually not sure. Maybe we will have to ask Gordon

---

_[test/src/hitbox_test.cpp, line 53 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1gbTrFOXYdwsQw9rn:-KS8sL1ET_ZezfxgvMPD:b-tatdaq) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/test/src/hitbox_test.cpp#L53)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> The tests are great, but can we organize them in a way that makes more sense?
> </details>
> OK, organized it to best effort haha

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68)_

**Comment by ExtraGravity** [comment]
Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_[src/hitbox.cpp, line 9 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1gsJ0K66MItsrcSYW:-KS9eZmk-2S5yUT0VtcA:b-xofwim) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.cpp#L9)):_
<details><summary>

<i>Previously, bseto (Byron Seto) wrote…</i></summary>



> OK
> </details>
> Did your astyle put the comma at the end? I like commas ate the beginning much better
> 
> ``` c++
>     : hitbox_type_(rect)
>     , hitbox_{x_pos, y_pos, width, height}
> ```

---

_[src/hitbox.h, line 22 at r4](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68#-KS1kJ5mds3uRcit2ngG:-KS9ubqG32FgVKUxFrxW:b-ravrxq) ([raw file](https://github.com/enochtsang/terminalfighter/blob/cb336736fcef211904ad8942460daaa49ec064ba/src/hitbox.h#L22)):_
<details><summary>

<i>Previously, enochtsang (Enoch Tsang) wrote…</i></summary>



> It's just an enum which evaluates to just an number in memory, so I think we can return pass by value for less complexity without losing performance.
> </details>
> Er I said pass by value, but this is ok too.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/enochtsang/terminalfighter/68)_

**Pull request merged**